### PR TITLE
ci: dockerize

### PR DESCRIPTION
**Pull request by gao-sun**
## Summary

- add `Dockerfile`
- add `docker-compose.yml` for quick demo
- read core launching params from env
- build, test, and upload docker image on master push

## Testing

- locally with `docker compose up`, works fine
- workflow [works fine](https://github.com/logto-io/logto/actions/runs/2542786904)

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+cache
+.*cache
+.DS_Store
+*.env
+.idea/
+*.pem
+.history
+.vscode
+.git
+.github
```
Comment (demonzoo) [inline .dockerignore:30]: nit: add `.gitignore`?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# Clean up
+RUN rm -rf .parcel-cache pnpm-*.yaml
+
+# Seal stage
+FROM node:16-alpine as app
+WORKDIR /etc/logto
+COPY --from=builder /etc/logto .
+EXPOSE 3001
+ENV NO_INQUIRY true
+ENTRYPOINT ["npm", "start"]
```
[the remark this answers is not in the document]
Comment (gao-sun) [reply Dockerfile:29]: no it;s for prod

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  app:
+    depends_on:
+      - postgres
+    image: logto:latest
+    ports:
+      - 3001:3001
+    environment:
+      ALL_YES: 1
+      NO_INQUIRY: 0
+      DB_URL_DEFAULT: postgres://postgres:p0stgr3s@postgres:5432
```
[the remark this answers is not in the document]
Comment (gao-sun) [reply docker-compose.yml:13]: no, and this env has another purpose

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Build stage
+FROM node:16-alpine as builder
+WORKDIR /etc/logto
+ENV CI=true
+COPY . .
```
Comment (wangsijie) [inline Dockerfile:5]: We can add `package.json` and `pnpm-lock.yaml` first, then it'll have 3 layers:
1. add package and lock file
2. pnpm i
3. add all other files

Then we can take advantage of docker layer cache.

But is not a blocker.

**Review by wangsijie — APPROVED** [review]

**Review by demonzoo — APPROVED** [review]

File: Dockerfile (hunk 4 above)
Comment (demonzoo) [reply Dockerfile:5]: Yes, this is preferred approach. 👍

Comment (gao-sun) [reply Dockerfile:5]: LOG-3197

Comment (gao-sun) [reply Dockerfile:5]: kind of tricky in monorepo, do it later

**Pull request merged**